### DISCUSSION TITLE
Understand ccache --show-stats output

### DISCUSSION THREAD

**Discussion by silverqx** [ccache/ccache#1455] [General]
This is some random output for the initial ccache build:

```
λ ccache --show-stats -v
Cache directory:    E:\tmp\ccache_tests
Config file:        E:\tmp\ccache_tests\ccache.conf
System config file: C:\ProgramData\ccache\ccache.conf
Stats updated:      Sat May 18 11:04:11 2024
Cacheable calls:    114 / 116 (98.28%)
  Hits:               0 / 114 ( 0.00%)
    Direct:           0
    Preprocessed:     0
  Misses:           114 / 114 (100.0%)
Uncacheable calls:    2 / 116 ( 1.72%)
  No input file:      2 /   2 (100.0%)
Successful lookups:
  Direct:             0 / 114 ( 0.00%)
  Preprocessed:       0 / 114 ( 0.00%)
Local storage:
  Cache size (GB):  0.1 / 0.7 (11.25%)
  Files:            228
  Hits:               0 / 114 ( 0.00%)
  Misses:           114 / 114 (100.0%)
  Reads:            228
  Writes:           228
```

I want to ask why all Files, Reads, and Writes are doubled? It's clear that every cached TU has 2 records, one record is the cached .obj file I suppose but what is the second record? 

If the CCACHE_DEBUG=1 is enabled then for every TU are logged these files `.ccache-input-c`, `.ccache-input-d`, `.ccache-input-text`, `.ccache-log`. Is the second record concatenation of all of these files? Or only some of them. I would like to fully understand this.

**Comment by silverqx** [comment]
If I enable the depend mode then reads **aren't** doubled, now I'm even more confused.

**Comment by jrosdahl** [comment]
By default, ccache writes two files to the cache on a cache miss:

1. A **manifest file**, which is stored in the cache using the _manifest key_. The manifest key is (slightly simplified) a hash of source code file + compiler options + environment info (e.g. compiler name, locale settings, current working directory) as the key. The manifest contains information about include file contents which are matched to a find a _result key_.
2. A **result file**, which is stored in the cache using the _result key_. The result key is (again, slightly simplified) found by hashing the output of running the preprocessor on the source code if it could not be found via the manifest. The result file contains all information needed to reproduce a compilation result. This includes the object file, but also other things like text sent to stdout/stderr and other files created by the compiler depending on compiler options, e.g. a dependency file (`.d`). See [this list](https://github.com/ccache/ccache/blob/d8523efc8328cc83b30ddeb3f89038bbd02c74a5/src/ccache/core/Result.hpp#L50-L94).

If the direct mode is disabled, manifest files are neither read nor written.

If the depend mode is enabled, the result key is not formed by running the preprocessor but by reading the dependency file created by the compiler. A manifest file is still used to map source code + compiler options + environment info to a result key.

See also [_How ccache works_](https://ccache.dev/manual/4.9.1.html#_how_ccache_works).

The `*.ccache-input-*` files created when debugging is enabled tell which data was hashed to form the keys used to look up the manifest and result files. See also [_Cache debugging_](https://ccache.dev/manual/4.9.1.html#_cache_debugging).

**Comment by silverqx** [comment]
I understand, I examined these manifest and results using `--inspect` and `--extract-result` yesterday and it's much more clear how it internally works.

> If I enable the depend mode then reads aren't doubled

I tried to restore with 100% cache hits now with the depend mode enabled and `reads` **are** doubled, I don't know what I saw 2 days ago 🙃 Maybe I was already too tired or I looked a line above or something.

Thx for explanation anyway
